iceoryx_hoofs/include/iceoryx_hoofs/cxx/span.hpp

```diff
@@ -0,0 +1,414 @@
+// Copyright (c) 2021 - 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#ifndef IOX_HOOFS_CXX_SPAN_HPP
+#define IOX_HOOFS_CXX_SPAN_HPP
+
+#include "iceoryx_hoofs/cxx/requires.hpp"
+#include "iceoryx_hoofs/cxx/span_iterator.hpp"
+#include "iceoryx_hoofs/cxx/type_traits.hpp"
+#include "iox/uninitialized_array.hpp"
+
+#include <algorithm>
+#include <cstdint>
+#include <cstdio>
+#include <limits>
+#include <utility>
+
+using namespace std;
+
+namespace iox
+{
+namespace cxx
+{
+// constants
+constexpr uint64_t dynamic_extent = std::numeric_limits<uint64_t>::max();
+
+template <class From, class To>
+constexpr bool is_convertible_v = std::is_convertible<From, To>::value;
+
+// Implementation of C++17 std::size() and std::data()
+template <class C>
+constexpr auto size(const C& c) -> decltype(c.size())
+{
+    return c.size();
```

Comment:
Function definition in header files is currently discouraged, should be moved to an inl file (only declared here).

IMO we have to move away from this as it leads to much boilerplate. The advantage is that the interface in the header is concise (but that is actually the task of doxygen).

Doxygen documentation is missing.

Reply:
Done :heavy_check_mark: